### DISCUSSION TITLE
can GaussianSource fourier transform be used for simply seperating TF/SF ?

### DISCUSSION THREAD

**Discussion by xianhuake** [NanoComp/meep#1931] [Q&A]
in the meep, when we seperate the total field and the scattering field, a normalization run involving an empty cell is conducted to save the incident fields from the source.
in order to save time, only on run is accepted to me. I found that the function fourier_transform of GaussianSource return the fourier transform of the cuttent evaluated at frequency. therefore, i wann to simply substract the source fourier in the total field to obtain the scattered field.
however, i has compared the resutlts return from src.fourier_transform and run involving an empty cell, a wavelength center shift is always ocurred, does the fourier transform not equal the results from a normalization run?
results are shown in the following figure, red line is the results of fourier_transform, and the green line corresponds to the run involving an empty cell.

![Figure_1](https://user-images.githubusercontent.com/20532980/152756033-813caaf8-cc2c-4ae8-b99f-b7ed58876a94.png)

**Comment by oskooi** [comment]
Note that the `fourier_transform` function of the `GaussianSource` class computes the Fourier transform of the *analytic* (i.e., non-discretized) Gaussian pulse. The discrepancy in your plot is likely due to the finite discretization. Try repeatedly doubling the `resolution` until the results have converged.

It would probably be useful to add an option to `fourier_transform` to return the Fourier transform of the discretized Gaussian pulse.

**Comment by xianhuake** [comment]
very thanks for your answer, i have do double the resolution in the grid volume, Unfortunately, the exactly same results are appeared.
the min wavelength is 0.3um and max wavelength is 0.8um, bloch boundary is applied in x and y direction, and pml in z, i change the resolution from 100 to 1000, noting changed, this means the results have converged.
![image](https://user-images.githubusercontent.com/20532980/153135365-b3a008bc-ce39-49bd-b0ca-a28b632c8604.png)

**Comment by oskooi** [comment]
Here is a script comparing the DFT (via `add_dft_fields`) and FT (via `fourier_transform`) of the E<sub>x</sub> fields in a 1d simulation. There is good agreement between the two methods. Note that the DFT fields are multiplied by a constant scaling factor to account for the finite discretization relative to the FT.

![dft_vs_ft_1d](https://user-images.githubusercontent.com/7152530/153536872-4471ad14-3c1f-4be2-9e9b-ad144265e422.png)

```py
import meep as mp
import numpy as np
import matplotlib
matplotlib.use('agg')
import matplotlib.pyplot as plt


resolution = 50

dpml = 1.0
sz = 10+2*dpml
cell_size = mp.Vector3(0,0,sz)
pml_layers = [mp.PML(dpml)]

fcen = 1.0
df = 0.2
src = mp.GaussianSource(fcen,fwidth=df)
sources = [mp.Source(src,
                     component=mp.Ex,
                     center=mp.Vector3(0,0,-0.5*sz+dpml),
                     size=mp.Vector3())]

sim = mp.Simulation(cell_size=cell_size,
                    boundary_layers=pml_layers,
                    sources=sources,
                    dimensions=1,
                    resolution=resolution)

nfreq = 51
mon = sim.add_dft_fields([mp.Ex],
                         fcen,
                         df,
                         nfreq,
                         center=mp.Vector3(0,0,0.5*sz-dpml),
                         size=mp.Vector3())

sim.run(until_after_sources=mp.stop_when_dft_decayed())

frqs = mp.get_flux_freqs(mon)
ex_dft = np.zeros(nfreq, dtype=np.complex128)
ex_ft = np.zeros(nfreq, dtype=np.complex128)
for n in range(nfreq):
    ex_dft[n] = sim.get_dft_array(mon, mp.Ex, n)
    ex_ft[n] = src.fourier_transform(frqs[n])

# constant scaling factor for DFT fields to account   
# for finite discretization relative FT           
scale_factor = abs(ex_ft[nfreq // 2]) / abs(ex_dft[nfreq // 2])

plt.figure()
plt.plot(frqs,scale_factor * np.abs(ex_dft),'bo-',label='DFT (scaled)')
plt.plot(frqs,np.abs(ex_ft),'ro-',label='FT')
plt.xlabel('frequency')
plt.ylabel('DFT or FT of Ex field (magnitude)')
plt.legend()
plt.title('1d cell with PMLs and GaussianSource')
plt.savefig('dft_vs_ft_1d.png',dpi=150,bbox_inches='tight')
```
